crates/event-streaming/proto/walrus/event/v1alpha/service.proto

```diff
@@ -0,0 +1,164 @@
+
+    // The method used to identify the event stream. The client MUST specify
+    // exactly one of these options. This makes the API explicit and avoids
+    // ambiguity between V1 (package-based) and V2 (object-based) streams.
+    oneof stream_selector {
+        // V1: Subscribe to all events emitted by a specific package.
+        PackageSubscription package_id = 2;
+
+        // V2: Subscribe to events managed by a dedicated stream head object.
+        StreamHead stream_head = 3;
+    }
+
+    // The desired proof types, in order of client preference.
+    // The server will attempt to provide the first type in this list that it
+    // supports for a given checkpoint. If empty, it uses its default behavior.
+    // Example: [MERKLE_PROOF_V2, CHECKPOINT_CONTENTS_V1] means client prefer V2,
+    // but will accept V1 if V2 is not available (would not accept response without
+    // a proof)
+    repeated ProofType desired_proof_types = 4;
+}
+
+// VerifiableEvents is the message type for the stream.
+// Each message represents a single checkpoint that contained Walrus events.
+message VerifiableEvents {
+    // The summary of the checkpoint containing these events. The client uses
+    // the `digest` field from this summary to verify the accompanying proof.
+    optional sui.rpc.v2beta.CheckpointSummary summary = 1;
+
+    // The list of all events found within this checkpoint.
+    repeated sui.rpc.v2beta.Event events = 2;
+
+    optional uint64 checkpoint_sequence_number = 3;
+
+    // The cryptographic proof that ties the events to the checkpoint summary.
+    // This is designed to be upgradeable.
+    optional Proof proof = 4;
+}
+
+message StreamHead {
+    // The Object ID of the dedicated stream head object.
+    optional string object_id = 1;
+}
+
+message PackageSubscription {
+    // The package object ID to subscribe to.
+    optional string package_id = 1;
+}
+
+// ============== Proof Structures ============================
+
+// Proof is an enum that allows the server to send different types of proofs.
+// This enables a future protocol upgrade from the simple V1 proof to the
+// more efficient V2 Merkle proof without breaking the API contract.
+message Proof {
+    oneof proof_type {
+        // V1 Proof: The entire raw contents of the checkpoint.
+        // The client verifies this by hashing the bcs and comparing
+        // it to the digest in the CheckpointSummary. And comparing the
+        // events against the content. Inefficient but simple.
+        sui.rpc.v2beta.CheckpointContents checkpoint_contents = 1;
+
+        // V2 Proof: A compact Merkle proof.
+        MerkleInclusionProof merkle_proof = 2;
+
+        // No-Proof Option: Sent when the client trusts the server and opts out
+        // of verification to save bandwidth and CPU.
+        google.protobuf.Empty proof_none = 3;
+    }
+}
+
+// Represents a Merkle inclusion proof for a checkpoint artifact.
+message MerkleInclusionProof {
```

Comment:
Is it useful to go into this level of detail instead of just setting `MerkleInclusionProof` and `MerkleNonInclusionProof` as bytes? I'm mostly worried about keeping the defs flexible (until the Merkle PR merges). I don't know anything about gRPC and its best practices though, so perhaps there are good reasons to do so...